added configuration best practices page for Kubernetes system components (#212)
cncf/cnf-wg: user-stories/configuration-best-practices.md

# Configuration best practices of Kubernetes system components

## Kubernetes API server security best practices!

### Disabling anonymous requests 

> Kubernetes API server can be configured to accept requests from unauthenticated sources. Anonymous requests have a username of system:anonymous, and a group name of system:unauthenticated and they can be assigned authorizations. 

Reviewer: Can we change this to such that it is only discussing disabling anonymous and unauthorized access?  The default Kubernetes API server configuration permits such access -- it doesn't have to be configured.
```suggestion
Kubernetes API server must be configured to reject requests from unauthenticated/anonymous sources; the default configuration allows such requests. Only authenticated users should be able to make requests to the API server. Set `--anonymous-auth` to `false`.
```

> Kubernetes API server must be configured to reject requests from unauthenticated/anonymous sources; the default configuration allows such requests. Only authenticated users should be able to make requests to the API server. Set `--anonymous-auth` to `false`.

Reviewer: minor typo
```suggestion
The Kubernetes API server must be configured to reject requests from unauthenticated/anonymous sources; the default configuration allows such requests. Only authenticated users should be able to make requests to the API server. Set `--anonymous-auth` to `false`.
```

This is a bad practice and only authenticated users should be able to make requests to the API server. Set `--anonymous-auth` to `false`.
### Enabling audit logging 

> API server audit logging is the functionality which enables operators to keep a record of external requests to the cluster. It records the requests together with the initiator identity therefore enabling security teams to track events happening in the system.

Reviewer: ```suggestion
API server audit logging is the functionality that enables operators to keep a record of requests to the cluster. It records the requests together with the initiator identity, therefore, enabling security teams to track events happening in the system.
```

Author: Valid point!

> API server configuration enables versatile configuration of audit logs which can both be delivered to a file or a webhook. See `--audit-log-path` or `--audit-policy-file` configuration.

Reviewer: ```suggestion
API server configuration enables versatile configuration of audit logs which can be delivered to a file or a webhook. See `--audit-log-path` or `--audit-policy-file` configuration.
```

### API authorization configuration
Kubernetes API server supports multiple authorization plug-ins:

> * Attribute-Based Access Control (ABAC) mode allows you to configure policies using local files.

Author: Yes, but the point here is rather the ability to manage access control. With ABAC you need to manage configuration file(s) as opposed to RBAC where you manage it through API.

* Node authorization is a special-purpose authorization mode that specifically authorizes API requests made by kubelets.
* AlwaysAllow allows all requests to be accepted (practically bypass access control)
AlwaysAllow is a bad choice.
The recommended way is to use “Role-based access control” since it gives a rich and manageable way to define access control in Kubernetes. Beyond the security perspective, this also makes the cluster compatible with applications requiring special authorization setup.

### Extending API access control limits with admission controllers
Kubernetes role based access control has its limitations. For example, if an entity has rights to create PODs it eventually gets full control over nodes running the cluster. This happens due to the fact that RBAC does not differentiate between a common POD and a privileged POD which has direct access to the host resources.
Admission controllers are optional components in a cluster enabling a more complex limitation strategy over the cluster. They are important security tool, especially in multi user environments.
### API server client authentication

> API server supports multiple client authentication technologies. These technologies answer different authentication needs and therefore there is not a single choice which can be considered a “best practice”, however there are best practices which can be applied to multiple choices and paramount the security of the API server and to the cluster in general.

Reviewer: is there a verb missing in "... and paramount the security of the API server"

Author: I am fixing these sentences

#### Anonymous users

> Allowing anonymous (unauthenticated users) is a bad practice in any production environment. Make sure `--anonymous-auth` is set to false in the API server configuration.

Reviewer 2: Dupliocate of 'disabling anonymous requests'?

Author: It is, I wanted to make a reference to this issue in the context of authentication settings. We can either:
* remove the sentence and just put a reference here
* remove the "disabling anonymous requests" paragraph and move it to here completely

#### Authentication credentials confidentiality
All authentication methods require some kind of client credentials to authenticate against. Known credential are:
Private key (client certificate authentication)
Tokens 
Passwords

These credentials need to be kept private in every case since they are the single factor for an attacker to access Kubernetes API resources.
Revocation and rotation of authentication credentials

> An operator must be able to manage authentication credentials and be able to revoke or rotate them. This is a must since different human errors can lead to leaks or theft, or simply access needs to be revoked.

Author: I am not sure about your point here. Do you mean that the most important that the API server has no access to the client secret?

> An operator must be able to manage authentication credentials and be able to revoke or rotate them. This is a must since different human errors can lead to leaks or theft, or simply access needs to be revoked.

Reviewer 3: @iawells can you clarify? Do you have a suggested change?

> An operator must be able to manage authentication credentials and be able to revoke or rotate them. This is a must since different human errors can lead to leaks or theft, or simply access needs to be revoked.

Reviewer: This statement is about revoking and rotating credentials -- both capabilities are needed and so "or" should be replaced by "and". Secondly, the statement after "since" is not really a justification for the need to rotate and revoke credentials; human errors can occur even if the best practices for revocation and rotation of credentials are followed.
```suggestion
An operator must be able to manage authentication credentials and be able to revoke and rotate them. This is a must best security practice. Rotation reduces the risk of misuse of long unchanged credentials (become widely available) while revocation of credentials needs to be performed for users who no longer need them,
```

> Different authentication methods support this functionality at different levels. Static token authentication requires restarts to the API server every time a token is revoked. Client certificate authentication revocation is not supported, but client certificates can have limited expiration periods therefore they can be “implicitly revoked” after some time and can be rotated regularly. The re-generation of newly signed client certificates makes operations more complex therefore operators tend to give longer expiration dates, which is not a good practice.

Reviewer: does this belong under "API server client authentication" above?

Author: Well, it is under the section "API server client authentication"

API server configuration enables versatile configuration of audit logs which can be delivered to a file or a webhook. See `--audit-log-path` or `--audit-policy-file` configuration.
### API authorization configuration
Kubernetes API server supports multiple authorization plug-ins:

> * Attribute-Based Access Control (ABAC) mode allows you to configure policies per user. The configuration is done using local files of the API server. This means that in order to change these policies, the operator needs to be able to access these files.

Reviewer: Whether the ABAC configuration uses local files or non-local files is an implementation detail. Are there specific access rights needed for these configuration files?

Reviewer: @slashben Thanks -- then please state that. For example,
Attribute-Based Access Control (ABAC) allows you to configure fine-grained access control policies per user, environment or resource attributes. The ABAC policies are specified in files; can't be managed using APIs. This means that in order to change these policies, the operator needs to be able to access these files.
